rfcs/0005-range-queries.md

```diff
@@ -0,0 +1,259 @@
+        T: RangeBounds<Bytes> 
+    {
+	    ...
+    }
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes tables in memory which which have not been flushed and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables. The latest records will have precedence following
+the same rules that a normal `get` query follows.
+
+We will rely on the latest checkpoint state in order to obtain references to memtables, SSTs,
+and sorted runs needed for the query. The `DbIterator` will expose the UUID of the checkpoint
+that it was generated from. The `DbIterator` will only maintain references to
+tables which cover the range specified in the `scan` parameters.
+
+Note that if the `DbReader` was created with a reference to a specific checkpoint,
+then the `DbIterator` will only refer to that checkpoint.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. We will build an iterator which sorts
+and merges memtables (e.g. with `MemTableIterator`), L0 tables (`SstIterator`),
+and sorted runs (`SortedRunIterator`). 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the query began scanning its range of records. For example,
+a query may include only a subset of records that were inserted in the same 
+`WriteBatch`. In other words, range queries will not have snapshot isolation 
+initially. 
+
+When a `Db` or `DbReader` is closed, active iterators will be invalidated. This
+will cause calls to `DbIterator::next` or `DbIterator::seek` to fail with an
+`InvalidatedIterator` error which indicates that the database was shutdown.
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-0004(https://github.com/slatedb/slatedb/blob/main/rfcs/0004-checkpoints.md)
```

Comment:
nit: typo in link *RFC-0004*

Reply:
Fixed! 56cf667462ba771f5e3a4ec083cdeda80df5a361